probe-rs/src/rtt.rs

```diff
@@ -115,27 +115,45 @@ impl Rtt {
 
     // Minimum size of the ControlBlock struct in target memory in bytes with empty arrays
     const MIN_SIZE: usize = Self::O_CHANNEL_ARRAYS;
+    const MIN_SIZE_64: usize = Self::O_CHANNEL_ARRAYS_64;
 
     // Offsets of fields in target memory in bytes
     const O_ID: usize = 0;
     const O_MAX_UP_CHANNELS: usize = 16;
     const O_MAX_DOWN_CHANNELS: usize = 20;
+    const O_MAX_DOWN_CHANNELS_64: usize = 24;
     const O_CHANNEL_ARRAYS: usize = 24;
+    const O_CHANNEL_ARRAYS_64: usize = 32;
 
     fn from(
         core: &mut Core,
         memory_map: &[MemoryRegion],
         // Pointer from which to scan
-        ptr: u32,
+        ptr: u64,
         // Memory contents read in advance, starting from ptr
         mem_in: Option<&[u8]>,
     ) -> Result<Option<Rtt>, Error> {
+        let is_64bit = match core.instruction_set()? {
+            probe_rs_target::InstructionSet::Thumb2
+            | probe_rs_target::InstructionSet::A32
+            | probe_rs_target::InstructionSet::RV32
+            | probe_rs_target::InstructionSet::RV32C
+            | probe_rs_target::InstructionSet::Xtensa => false,
+            probe_rs_target::InstructionSet::A64 => true,
+        };
         let mut mem = match mem_in {
             Some(mem) => Cow::Borrowed(mem),
             None => {
                 // If memory wasn't passed in, read the minimum header size
-                let mut mem = vec![0u8; Self::MIN_SIZE];
-                core.read(ptr.into(), &mut mem)?;
+                let mut mem = vec![
+                    0u8;
+                    if is_64bit {
```

Comment:
Could you instead store the 64bit-dependent constants in fields first, to avoid a bunch of (mostly) unnecessary if-else?

In other places, you could probably abstract away the `if 64-bit { do64 } else { do32 }` code into helper functions just to clean the implementation up a bit. I'm not sure what would be the nicest, but I'm sure we can do better than an if-else everywhere.

Reply:
> I'm sure we can do better than an if-else everywhere.

I agree.
I am refactoring RTT codes, but it will take time to complete.

https://github.com/tnishinaga/probe-rs/commit/72758b0fa7d59bf0be1b99902a377b0cc381be81

So, I would be happy if you could merge the current code first.

Reply:
Thank you for waiting.
The refactor is complete now.
Could you please review it again?